src/webgpu/api/operation/queue/writeBuffer.spec.ts

```diff
@@ -1,11 +1,228 @@
+    this.expectContents(buffer, expectedData);
+  }
+}
+
+export const g = makeTestGroup(F);
+
+const kTestData = range<number>(16, i => i);
+
+g.test('array_types')
+  .desc('Tests that writeBuffer correctly handles different TypedArrays and ArrayBuffer.')
+  .cases(params().combine(poptions('arrayType', kTypedArrays)).combine(pbool('useArrayBuffer')))
+  .fn(t => {
+    const { arrayType, useArrayBuffer } = t.params;
+
+    t.testWriteBuffer({ bufferOffset: 0, data: kTestData, arrayType, useArrayBuffer });
+  });
+
+g.test('multiple_writes_at_different_offsets_and_sizes')
+  .desc(
+    `
+Tests that writeBuffer currently handles different offsets and writes. This includes:
+- Non-overlapping TypedArrays and ArrayLists
+- Overlapping TypedArrays and ArrayLists
+- Writing zero data
+- Writing on zero sized buffers
+- Unaligned source
+- Multiple overlapping writes with decreasing sizes
+    `
+  )
+  .cases([
+    {
+      // Concatenate 2 Uint32Arrays
+      writes: [
+        {
+          bufferOffset: 0,
+          data: kTestData,
+          arrayType: 'Uint32Array',
+          useArrayBuffer: false,
+          dataOffset: 2,
+          dataSize: 2,
+        }, // [2, 3]
+        {
+          bufferOffset: 2 * Uint32Array.BYTES_PER_ELEMENT,
+          data: kTestData,
+          arrayType: 'Uint32Array',
+          useArrayBuffer: false,
+          dataOffset: 0,
+          dataSize: 2,
+        }, // [0, 1]
+      ], // Expected [2, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0]
+    },
+    {
+      // Concatenate 2 Uint8Arrays
+      writes: [
+        { bufferOffset: 0, data: [0, 1, 2, 3], arrayType: 'Uint8Array', useArrayBuffer: false },
+        { bufferOffset: 4, data: [4, 5, 6, 7], arrayType: 'Uint8Array', useArrayBuffer: false },
+      ], // Expected [0, 1, 2, 3, 4, 5, 6, 7]
+    },
+    {
+      // Overlap in the middle
+      writes: [
+        { bufferOffset: 0, data: kTestData, arrayType: 'Uint8Array', useArrayBuffer: false },
+        { bufferOffset: 4, data: [0], arrayType: 'Uint32Array', useArrayBuffer: false },
+      ], // Expected [0, 1, 2, 3, 0, 0 ,0 ,0, 8, 9, 10, 11, 12, 13, 14, 15]
+    },
+    {
+      // Overlapping arrayLists
+      writes: [
+        {
+          bufferOffset: 0,
+          data: kTestData,
+          arrayType: 'Uint32Array',
+          useArrayBuffer: true,
+          dataOffset: 2,
+          dataSize: 4 * Uint32Array.BYTES_PER_ELEMENT,
+        },
+        { bufferOffset: 4, data: [0x04030201], arrayType: 'Uint32Array', useArrayBuffer: true },
+      ], // Expected [0, 0, 1, 0, 1, 2, 3, 4, 0, 0, 3, 0, 0, 0, 4, 0]
+    },
+    {
+      // Write over with empty buffer
+      writes: [
+        { bufferOffset: 0, data: kTestData, arrayType: 'Uint8Array', useArrayBuffer: false },
+        { bufferOffset: 0, data: [], arrayType: 'Uint8Array', useArrayBuffer: false },
+      ], // Expected [0, 1, 2, 3, 4, 5 ,6 ,7, 8, 9, 10, 11, 12, 13, 14, 15]
+    },
+    {
+      // Zero buffer
```

Comment:
I'm not sure what this does, don't we want to do a single writebuffer of size 0 instead?